### DISCUSSION TITLE
"note on" when rotating a potentiometer

### DISCUSSION THREAD

**Discussion by RocktorT** [tttapa/Control-Surface#968] [General]
Hi!
How can you send a messages "note on" when rotating a potentiometer?

**Comment by tttapa** [comment]
Does this help? https://github.com/tttapa/Control-Surface/issues/161#issuecomment-609726411

**Reply by RocktorT** [reply]
No, I want that when the voltage on the analog input changes from 0 to Vcc, notes from C1 to C4 are sent sequentially with a constant velocity

**Reply by tttapa** [reply]
In that case you only need a minor change to the `ContinuousNoteSender::send` function: map the value to an offset in the range 0-36, and use `address + offset` as the address, and a constant value for the velocity.

**Reply by RocktorT** [reply]
thanks, I'll try to figure it out

**Comment by RocktorT** [comment]
when compiling #161 (comment) an error occurs:

C:\Users\user\AppData\Local\Temp\.arduinoIDE-unsaved2023930-4268-1h25ii0.sms3\sketch_oct30b\sketch_oct30b.ino:10:57: error: expected template-name before '<' token
 struct NotePotentiometer : MIDIFilteredAnalogAddressable<ContinuousNoteSender> {
                                                         ^
C:\Users\user\AppData\Local\Temp\.arduinoIDE-unsaved2023930-4268-1h25ii0.sms3\sketch_oct30b\sketch_oct30b.ino:10:57: error: expected '{' before '<' token
C:\Users\user\AppData\Local\Temp\.arduinoIDE-unsaved2023930-4268-1h25ii0.sms3\sketch_oct30b\sketch_oct30b.ino:10:57: error: expected unqualified-id before '<' token
C:\Users\user\AppData\Local\Temp\.arduinoIDE-unsaved2023930-4268-1h25ii0.sms3\sketch_oct30b\sketch_oct30b.ino:18:19: error: elements of array 'NotePotentiometer pots []' have incomplete type
 NotePotentiometer pots[] = {
                   ^~~~
C:\Users\user\AppData\Local\Temp\.arduinoIDE-unsaved2023930-4268-1h25ii0.sms3\sketch_oct30b\sketch_oct30b.ino:18:19: error: storage size of 'pots' isn't known

exit status 1

Compilation error: expected template-name before '<' token

**Reply by RocktorT** [reply]
what am I doing wrong?

**Comment by RocktorT** [comment]
I found a working sketch:
https://github.com/tttapa/Control-Surface/issues/893

**Reply by RocktorT** [reply]
I removed everything related to the encoder from there and left only one potentiometer.
Only here, when the pot is rotated, one note is sent with a variable velocity, but I need the note to change, but the velocity remains constant.

**Reply by RocktorT** [reply]
#include <Control_Surface.h> // Include the Control Surface library

struct ContinuousNoteSender {
    void send(uint8_t value, MIDIAddress address) {
        Control_Surface.sendNoteOn(address, value);
    }  
    static constexpr uint8_t precision() { return 7; }
};

struct NotePotentiometer : MIDIFilteredAnalog<ContinuousNoteSender> {
    NotePotentiometer(pin_t analogPin, MIDIAddress address)
      : MIDIFilteredAnalog(analogPin, address, {}) {}
};


USBMIDI_Interface midi;

using namespace MIDI_Notes;
NotePotentiometer pots[] = {
    {A0, note(C, 4)},
   
};



void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}

**Comment by RocktorT** [comment]
I did it!
#include "Control_Surface.h"  // Include the Control Surface library

struct ContinuousNoteSender {
  void send(uint8_t value, MIDIAddress address) {

      int delta = constrain(value, 0, 36);

      Control_Surface.sendNoteOn(address + delta, 100);
    //Serial.println(value);
  }
  static constexpr uint8_t precision() {
    return 7;
  }
};

struct NotePotentiometer : MIDIFilteredAnalog<ContinuousNoteSender> {
  NotePotentiometer(pin_t analogPin, MIDIAddress address)
    : MIDIFilteredAnalog(analogPin, address, {}) {}
};


USBMIDI_Interface midi;

using namespace MIDI_Notes;
NotePotentiometer pots[] = {
  { A0, note(C, 4) },
};



void setup() {
  Control_Surface.begin();  // Initialize Control Surface
  //Serial.begin(9600);
}

void loop() {
  Control_Surface.loop();  // Update the Control Surface
}

**Reply by RocktorT** [reply]
or rather, my son did)))
